src/main/java/com/minecolonies/coremod/colony/Structures.java

```diff
@@ -523,69 +541,276 @@ private static void checkDirectory(@NotNull final File directory)
+                    {
+                        md5Set.remove(schematicName);
+                        countInUseStructures++;
+                    }
+                }
+            }
+        }
+
+        //md5Set containd only the unused one
+        final Iterator<String> iterator = md5Set.iterator();
+        while (iterator.hasNext() && md5Set.size() + countInUseStructures >= Configurations.gameplay.maxCachedSchematics)
+        {
+            final StructureName sn = new StructureName(iterator.next());
+            if (deleteCachedStructure(sn))
+            {
+                iterator.remove();
+            }
+        }
+
+        return md5Set.size() + countInUseStructures < Configurations.gameplay.maxCachedSchematics;
+    }
+
+    /**
+     * add the md5 as a known structure in cache.
+     *
+     * @param md5 hash of the structure
+     */
+    public static void addMD5ToCache(@NotNull final String md5)
+    {
+        markDirty();
+        md5Map.put(Structures.SCHEMATICS_CACHE + SCHEMATICS_SEPARATOR + md5, md5);
+    }
+
+    /**
+     * get the set of cached schematic.
+     */
+    private static Set<String> getCachedMD5s()
+    {
+        final Set<String> md5Set = new HashSet<>();
+        for (final Map.Entry<String, String> md5 : md5Map.entrySet())
+        {
+            final StructureName sn = new StructureName(md5.getKey());
+            if (sn.getSection().equals(SCHEMATICS_CACHE))
+            {
+                md5Set.add(md5.getKey());
+            }
+        }
+        return md5Set;
+    }
+
+    /**
+     * delete a cached structure.
+     * delete the file and the md5 entry
+     *
+     * @param structureName the structure to delete
+     * @return True if the structure have been deleted, False otherwise
+     */
+    private static boolean deleteCachedStructure(@NotNull final StructureName structureName)
+    {
+        if (!SCHEMATICS_CACHE.equals(structureName.getPrefix()))
+        {
+            Log.getLogger().warn("Delete failed: Invalid name " + structureName);
+            return false;
+        }
+
+        if (!hasMD5(structureName))
+        {
+            Log.getLogger().warn("Delete failed: No MD5 hash found for " + structureName);
+            return false;
+        }
+
+        final File structureFile = MineColonies.proxy.getSchematicsFolder().toPath().resolve(structureName.toString() + SCHEMATIC_EXTENSION).toFile();
+        if (structureFile.delete())
+        {
+            md5Map.remove(structureName.toString());
+            return true;
+        }
+        else
+        {
+            Log.getLogger().warn("Failed to delete structure " + structureName);
+        }
+        return false;
+    }
+
+    /**
+     * mark Structures as dirty.
+     */
+    private static void markDirty()
+    {
+        dirty = true;
+    }
+
+    /**
+     * Class to handle schematic naming.
+     * It does extract information from a schematic using its name.
+     */
+    public static class StructureName
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce this class from 118 to the maximum allowed 30 or externalize it in a public class. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orionminecraft.com/coding_rules#rule_key=squid%3AS2972)